File: app/src/main/java/org/tvheadend/tvhclient/data/dao/ServerStatusDao.java
package org.tvheadend.tvhclient.data.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import org.tvheadend.tvhclient.data.entity.ServerStatus;

@Dao
public interface ServerStatusDao {

    @Query("DELETE FROM server_status WHERE connection_id = :id")
    void deleteByConnectionId(int id);

    @Query("SELECT s.*, " +
            "c.name AS connection_name " +
            "FROM server_status AS s " +
            "LEFT JOIN connections AS c ON c.id = s.connection_id AND c.active = 1")
    ServerStatus loadServerStatusSync();

    @Query("SELECT s.*, " +
            "c.name AS connection_name " +
            "FROM server_status AS s " +
            "LEFT JOIN connections AS c ON c.id = s.connection_id AND c.active = 1")
    LiveData<ServerStatus> loadServerStatus();

    @Insert
    void insert(ServerStatus serverStatus);

    @Update
    void update(ServerStatus serverStatus);

    @Delete
    void delete(ServerStatus serverStatus);

    @Query("DELETE FROM server_status")
    void deleteAll();

    @Query("SELECT s.*, " +
            "c.name AS connection_name " +
            "FROM server_status AS s " +
            "LEFT JOIN connections AS c ON c.id = s.connection_id " +
            "WHERE s.connection_id = :id")
    ServerStatus loadServerStatusByIdSync(int id);

    @Query("SELECT s.*, " +
            "c.name AS connection_name " +
            "FROM server_status AS s " +
            "LEFT JOIN connections AS c ON c.id = s.connection_id " +
            "WHERE s.connection_id = :id")
    LiveData<ServerStatus> loadServerStatusById(int id);
}

File: app/src/main/java/org/tvheadend/tvhclient/data/dao/ServerStatusDao.kt
package org.tvheadend.tvhclient.data.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import org.tvheadend.tvhclient.domain.entity.ServerStatus

@Dao
interface ServerStatusDao {

    @Query("DELETE FROM server_status WHERE connection_id = :id")
    fun deleteByConnectionId(id: Int)

    @Transaction
    @Query("$SERVER_STATUS_BASE_QUERY WHERE $CONNECTION_IS_ACTIVE")
    fun loadAllServerStatus(): LiveData<List<ServerStatus>>

    @Query("$SERVER_STATUS_BASE_QUERY WHERE $CONNECTION_IS_ACTIVE")
    fun loadActiveServerStatusSync(): ServerStatus

    @Query("$SERVER_STATUS_BASE_QUERY WHERE $CONNECTION_IS_ACTIVE")
    fun loadActiveServerStatus(): LiveData<ServerStatus>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(serverStatus: ServerStatus)

    @Update
    fun update(serverStatus: ServerStatus)

    @Delete
    fun delete(serverStatus: ServerStatus)

    @Query("DELETE FROM server_status")
    fun deleteAll()

    @Query("$SERVER_STATUS_BASE_QUERY WHERE s.connection_id = :id")
    fun loadServerStatusByIdSync(id: Int): ServerStatus

    @Query("$SERVER_STATUS_BASE_QUERY WHERE s.connection_id = :id")
    fun loadServerStatusById(id: Int): LiveData<ServerStatus>

    @get:Query("SELECT COUNT (*) FROM server_status AS s WHERE $CONNECTION_IS_ACTIVE")
    val serverStatusCount: LiveData<Int>

    companion object {

        const val SERVER_STATUS_BASE_QUERY = "SELECT DISTINCT s.* FROM server_status AS s "

        const val CONNECTION_IS_ACTIVE = " s.connection_id IN (SELECT id FROM connections WHERE active = 1) "
    }
}
